03-pipefilter.md

```diff
@@ -203,90 +177,285 @@ or copy data from one program to the other in memory,
-do something with what they've read,
-and write to standard output.
-
-The key is that any program that reads lines of text from standard input
-and writes lines of text to standard output
-can be combined with every other program that behaves this way as well.
-You can *and should* write your programs this way
-so that you and other people can put those programs into pipes to multiply their power.
+Another question that we can ask is
+"What is the smalest molecule inside `molecules`?".
+Since at `.pdb` files we have one line for each atom
+we can answer the question by
+(1) getting the number of lines of each file and
+(2) sorting the result.
+We already learned how to archive (1),
+we can use `wc -l *.pdb`.
+
+~~~ {.bash}
+$ wc -l *.pdb
+~~~
+~~~ {.output}
+   30 aldrin.pdb
+    7 ammonia.pdb
+   24 ascorbic-acid.pdb
+   18 benzaldehyde.pdb
+   30 camphene.pdb
+   78 cholesterol.pdb
+   22 cinnamaldehyde.pdb
+   33 citronellal.pdb
+   47 codeine.pdb
+   20 cubane.pdb
+   16 cyclobutane.pdb
+   23 cyclohexanol.pdb
+   13 cyclopropane.pdb
+   12 ethane.pdb
+   13 ethanol.pdb
+   54 ethylcyclohexane.pdb
+   14 glycol.pdb
+   79 heme.pdb
+   15 lactic-acid.pdb
+   49 lactose.pdb
+  248 lanoxin.pdb
+   53 lsd.pdb
+   49 maltose.pdb
+   35 menthol.pdb
+    9 methane.pdb
+   10 methanol.pdb
+   36 mint.pdb
+   44 morphine.pdb
+   41 mustard.pdb
+   33 nerol.pdb
+   51 norethindrone.pdb
+   30 octane.pdb
+   21 pentane.pdb
+   43 piperine.pdb
+   15 propane.pdb
+   25 pyridoxal.pdb
+   52 quinine.pdb
+   51 strychnine.pdb
+   20 styrene.pdb
+   49 sucrose.pdb
+   53 testosterone.pdb
+   42 thiamine.pdb
+   25 tnt.pdb
+   54 tuberin.pdb
+   34 tyrian-purple.pdb
+   23 vanillin.pdb
+   10 vinyl-chloride.pdb
+   55 vitamin-a.pdb
+ 1808 total
+~~~
+
+Today is our luck day
+and the shell has a command called `sort` that sort the lines of one file.
+Using the pipe we have
+
+~~~ {.bash}
+$ wc -l *.pdb | sort
+~~~
+~~~ {.output}
+   10 methanol.pdb
+   10 vinyl-chloride.pdb
+   12 ethane.pdb
+   13 cyclopropane.pdb
+   13 ethanol.pdb
+   14 glycol.pdb
+   15 lactic-acid.pdb
+   15 propane.pdb
+   16 cyclobutane.pdb
+ 1808 total
+   18 benzaldehyde.pdb
+   20 cubane.pdb
+   20 styrene.pdb
+   21 pentane.pdb
+   22 cinnamaldehyde.pdb
+   23 cyclohexanol.pdb
+   23 vanillin.pdb
+  248 lanoxin.pdb
+   24 ascorbic-acid.pdb
+   25 pyridoxal.pdb
+   25 tnt.pdb
+   30 aldrin.pdb
+   30 camphene.pdb
+   30 octane.pdb
+   33 citronellal.pdb
+   33 nerol.pdb
+   34 tyrian-purple.pdb
+   35 menthol.pdb
+   36 mint.pdb
+   41 mustard.pdb
+   42 thiamine.pdb
+   43 piperine.pdb
+   44 morphine.pdb
+   47 codeine.pdb
+   49 lactose.pdb
+   49 maltose.pdb
+   49 sucrose.pdb
+   51 norethindrone.pdb
+   51 strychnine.pdb
+   52 quinine.pdb
+   53 lsd.pdb
+   53 testosterone.pdb
+   54 ethylcyclohexane.pdb
+   54 tuberin.pdb
+   55 vitamin-a.pdb
+   78 cholesterol.pdb
+   79 heme.pdb
+    7 ammonia.pdb
+    9 methane.pdb
+~~~
+
+Ops. Something went catastrophic wrong.
```

Comment:
On my filesystem the sort operation works fine, due to the extra spaces the `wc` command puts in from of the numbers. The result is

```
$ wc -l *.pdb | sort
       7 ammonia.pdb
       9 methane.pdb
      10 methanol.pdb
      10 vinyl-chloride.pdb
      12 ethane.pdb
....
      78 cholesterol.pdb
      79 heme.pdb
     248 lanoxin.pdb
    1808 total
```

 This has been a thing I didn't like about the lesson before, and it hops over the need for `sort -n` when sorting numbers...